Set URL for videos to canonical YouTube video URL

robertknight opened the pull request:
- Add a `<link rel=canonical>` to the video transcript page, which the Hypothesis client will use as the URL for annotations.

 - For YouTube videos, set this URL to `https://www.youtube.com/watch?v={video_id}`.  This is the URL that the various other URLs (`https://youtube.com?v={video_id}`, `https://youtu.be/{video_id}`, `https://youtube.com/v/{video_id}`) ultimately redirect to and is what is used as the `<link rel=canonical>` on YouTube video pages.

Part of https://github.com/hypothesis/via/issues/941

----

**Testing:**

1. Go to http://localhost:9083/https://www.youtube.com/watch?v=x8TO-nrUtSI or any of the other recognized YouTube URLs (eg. http://localhost:9083/https://youtu.be/x8TO-nrUtSI)
2. Click the "?" button at the top of the client sidebar and go to "About this version"
3. The document URL displayed in the client should be "https://www.youtube.com/watch?v=x8TO-nrUtSI". There should be a `<link rel=canonical>` tag in the inner iframe's `<head>` with this same value.

If the URL contains additional parameters, such as `t={seconds}` to set the start time, these should *not* appear in the canonical URL.

robertknight committed:
Set URL for videos to canonical YouTube video URL

 - Add a `<link rel=canonical>` to the video transcript page, which the
   Hypothesis client will use as the URL for annotations.

 - For YouTube videos, set this URL to
   `https://www.youtube.com/watch?v={video_id}`.  This is the URL that the
   various other URLs (`https://youtube.com?v={video_id}`,
   `https://youtu.be/{video_id}`, `https://youtube.com/v/{video_id}`) ultimately
   redirect to and is what is used as the `<link rel=canonical>` on YouTube
   video pages.

Part of https://github.com/hypothesis/via/issues/941

seanh commented on tests/unit/via/views/view_video_test.py:
```diff
@@ -23,11 +23,15 @@ def test_it(
         video_url,
         ViaSecurityPolicy,
     ):
+        youtube_service.canonical_video_url.return_value = sentinel.canonical_video_url
```
You probably don't need to do this, you can just refer to `youtube_service.canonical_video_url.return_value` in the code below. Mocks already return unique values out of the box so unless the sentinel name is really adding anything to the code it's unnecessary

robertknight replied:
I followed the pattern of the existing code here (see `sentinel.youtube_video_id` below), although it did look non-idiomatic to me. Is there a reason why the video ID has to be done that way that I've overlooked?

seanh replied:
Yeah, just try removing it and see what breaks, it's [set to `None` by default](https://github.com/hypothesis/via/blob/a754bdef4f83ab9f924b991209fb19849a1e0962/tests/unit/services.py#L67). It probably shouldn't be set to `None` by default though as that's a bit surprising/unexpected: whatever tests need it to be `None` should probably set that in those particular tests

robertknight replied:
> Yeah, just try removing it and see what breaks,

I did this. It looks like this was added to the mocks so that the default behavior of any tests using it are that a URL is not a YouTube URL. Without this, a bunch of other tests started failing and would have needed to set up an override.

I have removed just the sentinel added in this test, and added a note about why the other one still exists.

seanh commented on via/services/youtube.py:
```diff
@@ -11,6 +11,19 @@ def __init__(self, enabled: bool):
+        associated with.
+        """
+        if not video_id:
+            raise ValueError("Invalid video ID")
```
I wouldn't bother with this `if`: the `quote_plus()` call below will raise `TypeError` if `video_id` is `None` (and perhaps also if `video_id` is invalid in other ways?). You could catch that `TypeError` and re-raise it as a `ValueError` if you want but given that nothing catches either I wouldn't bother.

We don't generally add validation code like this to our Python methods that're just called by other Python methods of ours. If you're worried about invalid video IDs coming from user-supplied data the place to validate that is at the edge: in a webargs schema on whatever views accept a `video_id` param, schema validation failures will trigger proper error responses

robertknight replied:
My thinking here was to make it slightly easier to debug if an empty string gets passed. I'm fine with just removing this though.

seanh replied:
Yeah, I'd say that empty-string video IDs shouldn't be getting into the code in the first place. If they are then we're missing some validation on some view(s)

robertknight replied:
OK, I have removed the `video_id` check here.

seanh commented on via/services/youtube.py:
```diff
@@ -11,6 +11,19 @@ def __init__(self, enabled: bool):
     def enabled(self):
         return self._enabled
 
+    def canonical_video_url(self, video_id: str) -> str:
+        """
+        Return the canonical URL for a YouTube video.
+
+        This is used as the URL which YouTube transcript annotations are
+        associated with.
+        """
```
Is this necessary?

I understand that `view_video.html.jinja2` needs to have the `<link rel="canonical" href="{{ video_url }}"/>` but I think you could just pass the non-normalized URL to the template, in `view_video.py::youtube()`:

```diff
diff --git a/via/views/view_video.py b/via/views/view_video.py
index 4c0e4b0..48770da 100644
--- a/via/views/view_video.py
+++ b/via/views/view_video.py
@@ -31,6 +31,7 @@ def youtube(request, url, **kwargs):
         "client_embed_url": request.registry.settings["client_embed_url"],
         "client_config": client_config,
         "video_id": video_id,
+        "video_url": url,
         "api": {
             "transcript": {
                 "doc": "Get the transcript of the current video",
```

I don't think it's Via's job to normalize / canonicalize YouTube URLs? Surely that's the job of the URL normalization and document equivalence code in h? Via isn't the only way that people might annotate YouTube videos and pages, after all.

robertknight replied:
> I don't think it's Via's job to normalize / canonicalize YouTube URLs?

For web pages, we do allow and encourage web page authors to indicate what the preferred version of a URL is using `<link rel=canonical>`, which originated as a hint for search engines, and YouTube does set this themselves. They also redirect the various URLs above to this canonical URL.

> Via isn't the only way that people might annotate YouTube videos and pages, after all.

When a user visits YouTube directly, YouTube includes this same `<link rel=canonical>` link on their pages.

seanh reviewed (commented):
I'm wondering if this should just use the non-normalized YouTube URL as the canonical URL, and do YouTube URL normalization with the rest of the URL normalization in h?

robertknight commented:
> I'm wondering if this should just use the non-normalized YouTube URL as the canonical URL, and do YouTube URL normalization with the rest of the URL normalization in h?

The URL normalization algorithm in h is effectively frozen at present. See https://github.com/hypothesis/h/issues/6552 (scroll down to bottom for my most recent thinking). I would like to resolve that, but for now I absolutely think using the canonical URL in the video player is the pragmatic option.

robertknight committed:
Remove check for falsey video ID, unnecessary use of `sentinel`

 - Remove check for invalid video ID, on the basis that these checks should be
   done earlier.

 - Remove unnecessary `canonical_video_url` sentinel. This was added to follow
   the pattern of the `get_video_id` sentinel, but that was needed because of
   the default configuration of the `get_video_id` mock. The `get_video_id`
   configuration is this way because it was a more convenient default for most
   tests.

seanh commented:
> > I'm wondering if this should just use the non-normalized YouTube URL as the canonical URL, and do YouTube URL normalization with the rest of the URL normalization in h?
> 
> The URL normalization algorithm in h is effectively frozen at present. See [hypothesis/h#6552](https://github.com/hypothesis/h/issues/6552) (scroll down to bottom for my most recent thinking). I would like to resolve that, but for now I absolutely think using the canonical URL in the video player is the pragmatic option.

Ok, this sounds fine to me :+1:
